cpp/include/cudf/io/csv.hpp

```diff
@@ -286,6 +288,11 @@ class csv_reader_options {
    */
   std::vector<int> const& get_infer_date_indexes() const { return _infer_date_indexes; }
 
+  /**
+   * @brief Returns per-column types.
+   */
+  std::vector<data_type> const& get_data_types() const { return _data_types; }
```

Comment:
I was thinking that we could use `std::variant` to store both options. This would allow us to overload the setters and expose a single getter that returns the variant - meaning that we keep the original API names after the old API is removed.
Also, I think it would be good to deprecate the old API, since we have a chance do to so.

Reply:
I have moved to unifying the two interfaces in a `std::variant` and overloading the `dtypes` accessors for `std::vector<data_type>` and `std::vector<std::string>`. 
This has the minor downside that the braced init list that we've previously used in `csv_test.cpp` for `dtypes` has become ambiguous for unary and binary init lists (because that's where we have overlap with `data_type`.
I.e.,
```cpp
cudf_io::csv_reader_options::builder(cudf_io::source_info{filepath})
      .dtypes({"date"})
...
```
Has now become explicit:

```cpp
cudf_io::csv_reader_options::builder(cudf_io::source_info{filepath})
      .dtypes(std::vector<std::string>{"date"})
...
```
I've done so far _all_ `.dtypes(...)` in `csv_test.cpp` for consistency, even for `n`-ary, with `n>2`. 
@vuule can you confirm that we want to roll with this?

Reply:
Had an offline discussion and decided to roll with this.